src/bst.py

```diff
@@ -123,6 +144,56 @@ def balance(self):
 
         return depth_left - depth_right
 
+    def _traverse(self, add, remove, size):
+        '''Traverse function
+            takes in other functions and a start_node'''
+        while size():
+            x = remove()
+            if x:
+                add(x.left, x.right)
+                yield x.value
+
+    def breadth_first_traversal(self):
+        '''perform a breadth first traversal, returns a list of
+           nodes in the graph
+        '''
+        q = Queue()
+        q.enqueue(self.root)
+
+        def add(a, b):
+            q.enqueue(a)
+            q.enqueue(b)
+
+        return self._traverse(add, q.dequeue, q.size)
+
+    def pre_order(self):
+        '''perform a breadth first traversal, returns a list of
+           nodes in the graph
+        '''
+        s = Stack()
+        s.push(self.root)
+
+        def add(a, b):
+            s.push(b)
+            s.push(a)
+        return self._traverse(add, s.pop, s.size)
+
+    def in_order(self):
+        '''Traverse function
+            takes in other functions and a start_node'''
+        if self.root:
+            return self.root.in_order()
+        else:
+            return []
```

Comment:
For these cases where the tree is empty, you are returning an empty list instead of a generator. This will probably work in most cases, however, the challenge for this assignment is to always return a generator. See if you can refactor so that `in_order` and `post_order` will _always_ return a generator, even when the tree is empty.